packages/framework/ai-collab/src/explicit-strategy/agentEditTypes.ts

```diff
@@ -0,0 +1,188 @@
+ * We could accomplish this via a path (probably JSON Pointer or JSONPath) from a possibly-null objectId, or wrap arrays in an identified object.
+ *
+ * TODO: only 100 object fields total are allowed by OpenAI right now, so larger schemas will fail faster if we have a bunch of schema types generated for type-specific edits.
+ *
+ * TODO: experiment using https://github.com/outlines-dev/outlines (and maybe a llama model) to avoid many of the annoyances of OpenAI's JSON Schema subset.
+ *
+ * TODO: without field count limits, we could generate a schema for valid paths from the root object to any field, but it's not clear how useful that would be.
+ *
+ * TODO: We don't supported nested arrays yet.
+ *
+ * TODO: Could omit edit contents for setRoot edits as the tree state is the result (or the other way around).
+ *
+ * TODO: Add a prompt suggestion API!
+ *
+ * TODO: Could encourage the model to output more technical explanations of the edits (e.g. "insert a new Foo after "Foo2").
+ *
+ * TODO: Get explanation strings from o1.
+ *
+ * TODO: Tests of range edits.
+ *
+ * TODO: SetRoot might be obscure enough to make the LLM avoid it. Maybe a general replace edit would be better.
+ *
+ * TODO: Handle 429 rate limit error in streamFromLlm.
+ *
+ * TODO: Add an app-specific guidance string.
+ *
+ * TODO: Give the model a final chance to evaluate the result.
+ *
+ * TODO: Separate system prompt into [system, user, system] for security.
+ *
+ * TODO: Top level arrays are not supported with current DSL.
+ *
+ * TODO: Structured Output fails when multiple schema types have the same first field name (e.g. id: sf.identifier on multiple types).
+ *
+ * TODO: Pass descriptions from schema metadata to the generated TS types that we put in the prompt
+ */
+
+/**
+ * TBD
+ */
+export const typeField = "__fluid_type";
+/**
+ * TBD
+ */
+export const objectIdKey = "__fluid_objectId";
+
+/**
+ * Describes an edit to a field within a node.
+ * @remarks what is the [key: string] for?
+ */
+export interface TreeEditObject {
+	[key: string]: TreeEditValue;
+	[typeField]: string;
+}
+/**
+ * An array of {@link TreeEditValue}'s, allowing a single TreeEdit to contain edits to multiple fields.
+ */
+export type TreeEditArray = TreeEditValue[];
+
+/**
+ * The potential values for a given TreeEdit. These values are typically a field within a node or an entire node,
+ * represented in JSON because they are expected to be returned by an LLM.
+ */
+export type TreeEditValue = JsonPrimitive | TreeEditObject | TreeEditArray;
+
+/**
+ * This is the the final object we expected from an LLM response.
+ * @remarks Because TreeEdit can be multiple different types (polymorphic),
+ * we need to wrap to avoid anyOf at the root level when generating the necessary JSON Schema.
+ */
+export interface EditWrapper {
+	// eslint-disable-next-line @rushstack/no-new-null
+	edit: TreeEdit | null;
```

Comment:
Why would this be null?

Reply:
We instruct the LLM to return null if it is done (no more edits to do to complete the task).

Reply:
If you allow the LLM to return an edit that doesn't do anything, would it be cleaner to have a `NoOp` type of `TreeEdit` rather than having this be nullable?

Reply:
I don't think so - the absence of an edit in the types seems a direct mirror of the absence of an edit in reality. Having an edit that is there, but does nothing, is an additional concept/possibility that we already have a way to express (undefined/null). I think the ideal would be for this to be `TreeEdit | undefined`, and we translate the undefined into null for JSON schema (since OpenAI JSON schema subset doesn't allow optionals).

Or, we could introduce some new edits like "MarkTaskCompleted" and "MarkTaskNotCompleted" and the LLM can return those as appropriate, instead of one of the existing tree edits. Though they don't quite fit into the structure here as-is. Maybe that would help the LLM perform better, who knows.

Reply:
I see, that makes sense. I forgot that this is the top level object since we don't allow it to give us multiple edits, in which case I would have expected an empty list of edits rather than edits that were null.